aten/src/ATen/cudnn/Handle.cpp

```diff
@@ -34,8 +35,67 @@ struct Handle {
 };
 
 std::mutex mutex;
-std::unordered_map<int, Handle> handles;
 
+// Handles are lazily created as different threads request them,
+// but are never destroyed until the end of the program.
+// This is the desired behavior: We want unused handles to stick around,
+// so they can be reused by other threads later.
+//
+// To prevent potential deadlocks, we explicitly choose not to cap the number
+// of handles that are created per device.
+// Example of danger: If we cap the max handles at 4, and 5 threads are sharing a device,
+// only 4 can make forward progress at any time. The other 4 will not release their
+// handles until they exit, so the fifth cannot make progress until then.  This is
+// not a problem...UNLESS all 5 threads attempt some sort of synchronization at an
+// intermediate point (ie, before any of them have exited).  We have no way to anticipate
+// or enforce that user threads will not attempt such intermediate synchronization.
+// The only way to ensure safety is to avoid imposing a cap on the number of handles.
+std::unordered_map<int, std::stack<Handle>> created_handles;
```

Comment:
Is there any reason we're using a `std::stack` backed by a `std::deque` here? I know in the stream pooling, we had to use a container that didn't move elements around because they were atomic. But I don't think that consideration applies here; these maps are just straight up protected by a mutex right.

Reply:
Yes, created_handles and available_handles are always protected by a mutex.  I used a stack because the original implementation used one, and it seemed like a natural data structure.

Reply:
Changed to std::vector in the latest update.